### ISSUE TITLE
kustomize plugin WASM support

### ISSUE THREAD

**Issue by mstrYoda** [kubernetes-sigs/kustomize#4956]
### Eschewed features

- [X] This issue is not requesting templating, unstuctured edits, build-time side-effects from args or env vars, or any other eschewed feature.

### What would you like to have added?

WASM runtime support for extensibility to kustomize

### Why is this needed?

The current plugin system depends on Go plugin which is gives limited usage area.

If kustomize support wasm plugin, people can write plugins in any language, version etc.

### Can you accomplish the motivating task without this feature, and if so, how?

Yes with Go plugin but a limited way.

### What other solutions have you considered?

Go plugin as mentioned.

### Feature ownership

- [X] I am interested in contributing this feature myself! 🎉

**Comment by natasha41575** [comment]
We like the idea of the feature, but before we can accept it, there are a few things we need to consider.

1. We would probably want to add WASM support as another KRM function runtime, in addition to docker and exec which we currently support. We need to think about how adding wasm runtime support will look, which should eventually become part of the[ KRM Function Specification](https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md). This specification is shared by multiple projects, including the kpt project which has already done some research into a wasm runtime and implemented an alpha version of it. We should try to leverage some of the work that has already been done there if we can. 

2. This will eventually be surfaced in kubectl, so we may eventually need to include it in an official KEP in k/enhancements repo. We already have a few KEPs there around KRM functions in kustomize so we could maybe add it to one of those.

3. There are some technical details around implementation, such as WASI vs wasmtime for the runtime that we need to discuss. We already have some small amount of research there, but need to discuss further.

There may be some other areas of discussion that we need before we move this forward. Because there is so much to discuss, I think it might be helpful if you can reach out to @KnVerey and I directly on slack. We will probably ask for a mini [kustomize KEP](https://github.com/kubernetes-sigs/kustomize/tree/master/proposals) but I think it will be helpful to sync on some of the details first. 

/triage under-consideration

**Comment by mstrYoda** [comment]
> We like the idea of the feature, but before we can accept it, there are a few things we need to consider.
> 
> 1. We would probably want to add WASM support as another KRM function runtime, in addition to docker and exec which we currently support. We need to think about how adding wasm runtime support will look, which should eventually become part of the[ KRM Function Specification](https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md). This specification is shared by multiple projects, including the kpt project which has already done some research into a wasm runtime and implemented an alpha version of it. We should try to leverage some of the work that has already been done there if we can.
> 2. This will eventually be surfaced in kubectl, so we may eventually need to include it in an official KEP in k/enhancements repo. We already have a few KEPs there around KRM functions in kustomize so we could maybe add it to one of those.
> 3. There are some technical details around implementation, such as WASI vs wasmtime for the runtime that we need to discuss. We already have some small amount of research there, but need to discuss further.
> 
> There may be some other areas of discussion that we need before we move this forward. Because there is so much to discuss, I think it might be helpful if you can reach out to @KnVerey and I directly on slack. We will probably ask for a mini [kustomize KEP](https://github.com/kubernetes-sigs/kustomize/tree/master/proposals) but I think it will be helpful to sync on some of the details first.
> 
> /triage under-consideration

Hi, thank you for your response. I will reach out to you on slack 👍
